### PR TITLE
Transaction equality, hashing, and attributes

### PR DESCRIPTION
**Pull request by carver**
Note that this is not a final draft ready for merging, it's meant as a reference implementation for discussion.

Some key discussion points:
1. Are my goals incompatible with other project goals?
2. Is there an alternative approach that satisfies my goals? 
3. If returning a new Tx object is acceptable, are there other important implementation issues to consider that I missed?

### What was wrong?

These were some key benefits I was looking for:
* equality test using the transaction hash value (For example, in my first several use cases, I wanted: `transaction_before_block_inclusion == transaction_after_block_inclusion`)
* can be used in a set, to prevent duplicate transactions with the same hash
* can access the transaction values as attributes (It just feels more natural to use `tx.hash` than `tx['hash']`)
* transaction object should be immutable, to eliminate a whole class of bugs at very low cost

In short, I want to use transactions as a first class object.

### How was it fixed?

I updated the output formatter to return an object of a new class that acts like a dictionary for a lot of backwards compatibility (except for equality tests, since changing that was an explicit goal).

#### Cute Animal Picture

Cute animal picture

### CODE REVIEW

File: web3/transaction.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
+
+
+class Transaction(object):
```
Comment (pipermerriam) [inline web3/transaction.py]: A few things.

1. Lets move this object to `web3.utils.datastructures`.
2. Rename to `AttributeDict` and remove any language that is transaction specific.
3. Remove the immutability functionality (will explain elsewhere)
3. Inherit the class from `collections.Mapping`

File: web3/transaction.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        @param values a dictionary of transaction values
+        '''
+        self.__dict__ = values
+        try:
+            self._unique_txid()
+        except AttributeError as e:
+            assert False, "Transaction must be uniquely defined. Failed because: %s" % e
+        self._locked = True
+
+    def _unique_txid(self):
```
Comment (pipermerriam) [inline web3/transaction.py]: Lets remove this until it's actually needed.

File: web3/transaction.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    def __setitem__(self, key, val):
+        raise NotImplementedError(Transaction.IMMUTABILITY_WARNING)
+
+    def __delitem__(self, key):
+        raise NotImplementedError(Transaction.IMMUTABILITY_WARNING)
+
+    def __iter__(self):
+        return iter(self.__dict__)
+
+    def todict(self):
```
Comment (pipermerriam) [inline web3/transaction.py]: I'd like to remove this and simply rely on `dict(txn_obj)` for casting to a dictionary.

File: web3/transaction.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        return dict(self.__dict__)
+
+    def items(self):
+        return self.__dict__.items()
+
+    def iteritems(self):
```
Comment (pipermerriam) [inline web3/transaction.py]: Is this still required?  I've been under the impression that for `2.7/3.4/3.5` that merely implementing `.items()` was sufficient.

Comment (carver) [inline web3/transaction.py]: This was actually from me poking around trying to get dict(txn_obj) working. It looks like the ideal solution is inheriting from `collections.Mapping`, so I'm happy to remove this.
